src/lib/litegraph/test/subgraph/NestedSubgraphSerialization.test.ts

```diff
@@ -0,0 +1,187 @@
+    })
+
+    const { node: kSamplerNode } = createNodeWithNumberWidget(
+      'KSampler',
+      32115495257102
+    )
+    innerSubgraph.add(kSamplerNode)
+    innerSubgraph.inputNode.slots[0].connect(
+      kSamplerNode.inputs[0],
+      kSamplerNode
+    )
+    innerSubgraph.outputNode.slots[0].connect(
+      kSamplerNode.outputs[0],
+      kSamplerNode
+    )
+
+    // Outer subgraph (subgraph 2) - contains the inner subgraph
+    const outerSubgraph = createTestSubgraph({
+      inputs: [{ name: 'seed', type: 'number' }],
+      outputs: [{ name: 'latent', type: 'LATENT' }]
+    })
+
+    const innerSubgraphNode = createTestSubgraphNode(innerSubgraph)
+    outerSubgraph.add(innerSubgraphNode)
+    outerSubgraph.inputNode.slots[0].connect(
+      innerSubgraphNode.inputs[0],
+      innerSubgraphNode
+    )
+    outerSubgraph.outputNode.slots[0].connect(
+      innerSubgraphNode.outputs[0],
+      innerSubgraphNode
+    )
+
+    // Main graph - contains the outer subgraph
+    const mainGraph = new LGraph()
+    const outerSubgraphNode = createTestSubgraphNode(outerSubgraph)
+    mainGraph.add(outerSubgraphNode)
+
+    // The outer subgraph node should have promoted the widget from the inner subgraph
+    expect(outerSubgraphNode.widgets).toHaveLength(1)
+    expect(outerSubgraphNode.serialize_widgets).toBe(true)
+
+    // Change the value in the outer subgraph node
+    const newSeedValue = 99999
+    outerSubgraphNode.widgets[0].value = newSeedValue
+
+    // Serialize the outer subgraph node
+    const serialized = outerSubgraphNode.serialize()
+
+    // The widget value should be preserved in serialization
+    expect(serialized.widgets_values).toBeDefined()
+    expect(serialized.widgets_values).toEqual([newSeedValue])
+  })
+
+  it('should disable widget serialization when all widgets are removed', () => {
+    const subgraph = createTestSubgraph({
+      inputs: [{ name: 'value', type: 'number' }]
+    })
+
+    const { node } = createNodeWithNumberWidget('Test Node')
+    subgraph.add(node)
+    subgraph.inputNode.slots[0].connect(node.inputs[0], node)
+
+    const subgraphNode = createTestSubgraphNode(subgraph)
+
+    // Initially should have widgets and serialization enabled
+    expect(subgraphNode.widgets).toHaveLength(1)
+    expect(subgraphNode.serialize_widgets).toBe(true)
+
+    // Remove the widget
+    subgraphNode.removeWidgetByName('value')
+
+    // Should disable serialization when no widgets remain
+    expect(subgraphNode.widgets).toHaveLength(0)
+    expect(subgraphNode.serialize_widgets).toBe(false)
+  })
+
+  it('should handle reconfiguration correctly with widget serialization', () => {
+    const subgraph = createTestSubgraph({
+      inputs: [{ name: 'value', type: 'number' }]
+    })
+
+    const { node } = createNodeWithNumberWidget('Test Node', 123)
+    subgraph.add(node)
+    subgraph.inputNode.slots[0].connect(node.inputs[0], node)
+
+    const subgraphNode = createTestSubgraphNode(subgraph)
```

Comment:
Some of these are repeated enough that I'd consider extracting a test utility. It'd clean up the tests and make clear which pieces differ in each Arrange.